jax/experimental/host_callback.py

```diff
@@ -0,0 +1,307 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of an experimental primitive for printing, including
+
+   from transformed and compiled code.
+
+See documentation for `id_print` below.
+For usage example, see tests/host_callback_test.py.
+
+Implementation plan:
+  * Write the API for the `id_print` primitive, using data-dependence as
+    explained in `id_print` documentation.
+  * Implement the transformations. DONE (except pmap)
+  * Implement the JIT for CPU using CustomCall in C++.
+    DONE (except unit tests do not run in OSS)
+  * Implement the JIT for GPU using also CustomCall in C++. STARTED.
+  * Explore how to pipe the printed data back to the Colab cell,
+    when running in Colab. ?
+  * Explore implementation using outfeed, hoping that it works for all
+    platforms, and can pipe data more easily. STARTED.
+  * Explore feeding the data back to the Python program (the `id_tap`
+    primitive). ?
+  * Explore a simpler API that uses Python program-order, instead of
+    data dependency-order. Need to add support to JAX for stateful primitives.
+
+"""
+
+from functools import partial
+import io
+
+from jax import lax, core
+from jax.lib import pytree, xla_bridge
+from jax.interpreters import ad, xla, batching
+from jax.interpreters import partial_eval as pe
+from jaxlib import xla_client
+from jaxlib import xla_extension
+
+import msgpack  # type: ignore
+import numpy as onp
+import os
+from typing import Any, Dict, List, Sequence, Tuple
+
+if os.getenv("JAX_ENABLE_JIT_PRINT", "false") != "false":
```

Comment:
This is a temporary flag to allow Travis to test the Python code, even before the C++ code made it into jaxlib.